src/mono/mono/mini/interp/interpgo.c

```diff
@@ -0,0 +1,395 @@
+	memcpy (table->data + table->size, hash, MM3_HASH_BYTE_SIZE);
+	table->size = required_size;
+
+	mono_os_mutex_unlock (&table->mutex);
+}
+
+static void
+table_sort_locked (interpgo_table *table) {
+	mono_qsort (table->data, table->size / MM3_HASH_BYTE_SIZE, MM3_HASH_BYTE_SIZE, hash_comparer);
+}
+
+static void
+compute_method_hash (MonoMethod *method, uint8_t outbuf[MM3_HASH_BYTE_SIZE]) {
+	// method token + image guid
+	size_t size = sizeof(uint32_t) + 16;
+	uint32_t *inbuf = alloca (size);
+	// method tokens are globally unique within a given assembly
+	inbuf[0] = mono_method_get_token (method);
+	// use the assembly guid as a unique id for the assembly
+	MonoImage *image = m_class_get_image (mono_method_get_class (method));
+	memcpy (inbuf + 1, mono_image_get_guid (image), 16);
+
+	MurmurHash3_128 (inbuf, size, 0x43219876, (uint8_t *)outbuf);
+}
+
+gboolean
+mono_interpgo_should_tier_method (MonoMethod *method) {
+	if (method->wrapper_type != MONO_WRAPPER_NONE)
+		return FALSE;
+
+	// If we didn't load a table, don't bother hashing the method.
+	if (!loaded_table)
+		return FALSE;
+
+	uint8_t hash[MM3_HASH_BYTE_SIZE];
+	compute_method_hash (method, hash);
+
+	if (table_lookup (loaded_table, hash)) {
+#if LOG_INTERPGO
+		char * name = mono_method_full_name (method, TRUE);
+		g_print ("Tiering %s because it was in the interpgo table\n", name);
+		g_free (name);
+#endif
+
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
+void
+mono_interpgo_method_was_tiered (MonoMethod *method) {
+	if (!mono_opt_interpgo_recording)
+		return;
+
+	if (method->wrapper_type != MONO_WRAPPER_NONE)
+		return;
+
+	uint8_t hash[MM3_HASH_BYTE_SIZE] = {0};
+	compute_method_hash (method, hash);
+	table_add (&building_table, hash);
+
+#if LOG_INTERPGO
+	char * name = mono_method_full_name (method, TRUE);
+	g_print ("added %s to table\n", name);
+	g_free (name);
+#endif
+}
+
+#if HOST_BROWSER
+
+#include <emscripten.h>
+
+// We disable this diagnostic because EMSCRIPTEN_KEEPALIVE makes it a false alarm, the keepalive
+//  functions are being used externally. Having a bunch of prototypes is pointless since these
+//  functions are not consumed by C anywhere else
+#pragma clang diagnostic ignored "-Wmissing-prototypes"
+
+EMSCRIPTEN_KEEPALIVE int
+mono_interpgo_load_table (uint8_t * data, int data_size) {
+	interpgo_table *result = g_malloc0 (sizeof (interpgo_table));
+	// The table storage format is [uint32 size] [data...]
+	uint32_t size = *(uint32_t *)data;
+
+	g_print ("Loading %d bytes of interpgo data (table size == %zu)\n", data_size, size);
+
+	result->data = g_malloc0 (data_size);
+	g_assert ((int64_t)size < (int64_t)data_size);
+	result->size = size;
+	result->capacity = data_size;
+	memcpy (result->data, data + sizeof (uint32_t), result->size);
+
+	// Atomically swap the new table in
+	interpgo_table *old_table = mono_atomic_xchg_ptr ((volatile gpointer*)&loaded_table, result);
+
+	if (old_table) {
```

Comment:
Under what circumstances do we expect the table to already be there? If another thread loads it?  Or do we get an updated table under some circumstances (maybe I missed where that happens)?

If it's a race with other threads, the usual lock-free approach would be a CAS with the expected value being NULL, and then dispose of the new table, not the old one.

Reply:
Yeah in multithreaded scenarios it's possible that multiple threads could try to load a table. We could make attempts after the first fail if we wanted.